crates/bitvm2/src/committee/api.rs

```diff
@@ -0,0 +1,339 @@
+use bitcoin::{Witness, key::Keypair, hex::FromHex, TapSighashType};
+use goat::transactions::signing_musig2::{generate_aggregated_nonce, generate_taproot_aggregated_signature};
+use musig2::{secp256k1::schnorr::Signature, AggNonce, PartialSignature, PubNonce, SecNonce};
+use sha2::{Sha256, Digest};
+use goat::transactions::{
+    base::BaseTransaction,
+    pre_signed::PreSignedTransaction,
+    pre_signed_musig2::get_nonce_message,
+    signing_musig2::generate_taproot_partial_signature,
+};
+use goat::connectors::{
+    connector_0::Connector0,
+    connector_5::Connector5,
+    connector_d::ConnectorD,
+};
+use crate::types::{Bitvm2Graph, Error};
+
+pub const COMMITTEE_PRE_SIGN_NUM: usize = 5;
+
+pub fn committee_pre_sign(
+    committee_member_keypair: Keypair,
+    committee_member_sec_nonce: [SecNonce; COMMITTEE_PRE_SIGN_NUM],
+    committee_agg_nonce: [AggNonce; COMMITTEE_PRE_SIGN_NUM],
+    graph: &Bitvm2Graph,
+) -> Result<[PartialSignature; COMMITTEE_PRE_SIGN_NUM], Error> {
+    let verifier_context = graph.parameters.get_verifier_context(committee_member_keypair);
+    let mut res: Vec<PartialSignature> = vec![];
+
+    {   // take-1 input-0, use nonce[0]
+        let tx = &graph.take1;
+        let input_index = 0;
+        let nonce_index = 0;
+        let sighash_type = TapSighashType::All;
+        match generate_taproot_partial_signature(
+            &verifier_context,
+            tx.tx(),
+            &committee_member_sec_nonce[nonce_index],
+            &committee_agg_nonce[nonce_index],
+            input_index,
+            tx.prev_outs(),
+            &tx.prev_scripts()[input_index],
+            sighash_type,
+        ) {
+            Ok(v) => res.push(v),
+            Err(e) => return Err(format!("fail to sign {} input-{input_index}: {e}", tx.name())),
```

Comment:
use anyhow::bail! to replace the `return Err(format!())`. https://docs.rs/anyhow/latest/anyhow/macro.bail.html

Reply:
We can use anyhow::Result as the return data, hence you can call like `generate_taproot_partial_signature(..)?;` directly.

Reply:
acknowledged, replaced return value accordingly